refactor(test): improve L1 files OPContractsManager and ProtocolVersions test coverage and quality with AI assistance [Prompt v0.7.3]

aliersh opened the pull request:
## Overview

This PR improves test coverage and quality for both `OPContractsManager` and `ProtocolVersions` using [Prompt v0.7.3](https://www.notion.so/Prompt-v0-7-3-241f153ee16280ca8282dba67e900268?pvs=21). This version builds on previous releases with stronger enforcement around **test intent**, **failure analysis**, **test uniqueness**, and **redundancy avoidance**.

Prompt v0.7.3 introduces key enhancements:

- Clear separation between **intent** and technical **side effects** in test outcomes  
- Elimination of **redundant tests** with different values testing the same logic  
- Detection and rejection of **duplicate getter tests** if logic is already verified elsewhere  
- Improved failure path handling and stronger guidance for safe test assumptions

These improvements help ensure tests verify meaningful, business-relevant behavior without inflating coverage for its own sake.

---

## OPContractsManager

### New Tests

- `test_l1ContractsRelease_rcTrue_succeeds`:  
  Validates that `l1ContractsRelease()` returns true under expected release conditions.

- `test_validate_succeeds`:  
  Verifies that `validate()` passes when called with valid parameters and deployed contract references.

- `test_validate_allowFailureFalse_reverts`:  
  Ensures `validate()` reverts when failure is not allowed and validation fails.

- `test_validateWithOverrides_succeeds`:  
  Confirms that `validateWithOverrides()` behaves correctly with default overrides.

- `testFuzz_validateWithOverrides_varyingOverrides_succeeds`:  
  Fuzz test exercising override behavior across a wide range of inputs.

- `test_blueprints_succeeds`:  
  Checks that blueprint lookups return correct addresses from the internal mapping.

- `test_implementations_succeeds`:  
  Confirms implementation mapping returns the correct contract addresses.

### Converted and Enhanced Tests

- `test_calculatesBatchInboxAddress_succeeds` →  
  `testFuzz_chainIdToBatchInboxAddress_varyingChainIds_succeeds`:  
  Adds fuzz testing to validate logic across a wide variety of chain IDs.

- `test_chainIdToBatchInboxAddress_knownValues_succeeds`:  
  Preserves targeted testing of known input-output mappings for completeness.

- `test_semver_works` → `test_version_succeeds`:  
  Renamed and enhanced with more precise assertions on version formatting.

### Other Improvements

- Validation-related tests now use actual deployed contract instances instead of placeholder or invalid addresses.

### Test Count

- 7 new tests added  
- 1 test converted into a fuzz test  
- 2 tests enhanced or renamed  
- All 33 tests pass

---

## ProtocolVersions

### Converted to Fuzz

- `testFuzz_setRequired_notOwner_reverts`:  
  Fuzz test confirming only the owner can call `setRequired()`.

- `testFuzz_setRecommended_notOwner_reverts`:  
  Same as above for the `setRecommended()` function.

### Organizational Improvements

- All test contracts now follow the source file’s function order.
- Getter logic is separated into dedicated test contracts for clarity.

### Test Count

- 2 tests converted into fuzz tests  
- All 8 tests pass

---

## Impact

- 7 new tests added across both files  
- 3 tests converted into fuzz tests  
- 2 tests renamed or refactored for precision  
- All tests aligned with Prompt v0.7.3 core principles:
  - No redundant tests
  - No tests that pass for unrelated technical reasons
  - No duplicate logic under different values
- Full CI pass and heavy fuzz validation (20,001 runs)

---

## Follow-up Fixes

### Revert Reason Fix (Semgrep Compliance)

One manual correction was made to resolve a Semgrep rule violation:

- **Issue**: The Semgrep rule `sol-safety-expectrevert-no-args` requires specifying the exact revert reason in `vm.expectRevert()`.
- **Fix**: In `test_validate_allowFailureFalse_reverts`, the line:
  ```solidity
  vm.expectRevert();
  was replaced with:
  vm.expectRevert(bytes("OPContractsManagerStandardValidator: PROXYA-10,SYSCON-40,L1xDM-20,L1SB-20,MERC20F-20,L721B-20,PORTAL-20,DF-20,DF-3
  0,PDDG-40,PDDG-DWETH-20,PDDG-DWETH-30,PDDG-DWETH-40,PDDG-ANCHORP-20,PDDG-VM-10,PDDG-130,PLDG-10,LOCKBOX-20"));
  ```
- Outcome: The test continues to pass and the Semgrep violation is resolved.

### CI Failures: Environment-Specific Getter Tests (ProtocolVersions)

The following test contracts were removed:

- `ProtocolVersions_Required_Test`
- `ProtocolVersions_Recommended_Test`

These tests failed in CI due to environment-dependent configuration values (e.g., protocol version values differing from local test
setup). They were also redundant:

- The `test_initialize_values_succeeds` test already verifies that both getters return the correct values after initialization.
- The getter functions are simple views that read directly from initialized state.
- Maintaining separate tests added fragility without increasing meaningful coverage.

### Library Usage Optimization (OPContractsManager)

Team review identified custom helper functions that duplicated existing library functionality:

- Issue: Custom `_contains` helper function and manual string slicing for `endsWith` functionality in OPContractsManager tests.
- Fix: Replaced with `LibString` library functions:
```solidity
// Before
assertTrue(_contains(version, "."), "version should contain dots for semver format");
assertEq(Bytes.slice(releaseBytes, releaseBytes.length - 3, 3), "-rc", "release should end with '-rc'");

// After  
import { LibString } from "@solady/utils/LibString.sol";
assertTrue(LibString.contains(version, "."), "version should contain dots for semver format");
assertTrue(LibString.endsWith(release, "-rc"), "release should end with '-rc'");
```

- Outcome: Removed 18-line custom helper function, improved code readability, leveraged battle-tested library functions, and achieved slight gas optimization.

aliersh committed:
refactor(test): improve OPContractsManager test coverage and quality

- add 7 tests for uncovered functions: l1ContractsRelease, validate, validateWithOverrides, blueprints, implementations
- convert 1 test to fuzz test: chainIdToBatchInboxAddress
- enhance version test with better assertions
- fix validation tests to use proper deployed contracts

aliersh committed:
refactor(test): improve ProtocolVersions test coverage and quality

- add 2 test contracts for getter functions
- convert 2 tests to fuzz tests for better coverage
- improve test organization following source function order

aliersh committed:
fix(test): specify revert reason in OPContractsManager validation test

- replace vm.expectRevert() with specific error message in test_validate_allowFailureFalse_reverts
- resolves semgrep violation sol-safety-expectrevert-no-args

aliersh committed:
fix(test): remove environment-dependent getter tests from ProtocolVersions

- remove ProtocolVersions_Required_Test contract
- remove ProtocolVersions_Recommended_Test contract

JosepBove commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -2031,8 +2045,32 @@ contract OPContractsManager_Version_Test is OPContractsManager_TestInit {
+    }
+
+    /// @notice Helper to check if a string contains a substring.
+    function _contains(string memory _str, string memory _substr) internal pure returns (bool) {
```
Not needed, we have the `LibString` library that has a member called `LibString.contains(subject, needle)`

aliersh replied:
Fixed in [1253c3b](https://github.com/ethereum-optimism/optimism/pull/16963/commits/1253c3bedfc5bbbf54516d21ed5af40f6200c478)

JosepBove commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -2083,3 +2121,134 @@ contract OPContractsManager_SetRC_Test is OPContractsManager_Upgrade_Harness {
+    /// @notice Tests that l1ContractsRelease returns the correct value when isRC is true.
+    function test_l1ContractsRelease_rcTrue_succeeds() public view {
+        // isRC is true by default
+        string memory release = opcm.l1ContractsRelease();
```
It can use `LibString` in this function too, we have `LibString.endsWith(subject, needle)`

aliersh replied:
Fixed in [1253c3b](https://github.com/ethereum-optimism/optimism/pull/16963/commits/1253c3bedfc5bbbf54516d21ed5af40f6200c478)

aliersh committed:
refactor(test): replace custom string helpers with LibString library

- replace custom _contains function with LibString.contains in version test
- replace Bytes.slice endsWith logic with LibString.endsWith in setRC test
- remove redundant _contains helper function
- add LibString import from @solady/utils/LibString.sol

smartcontracts commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -2083,3 +2101,134 @@ contract OPContractsManager_SetRC_Test is OPContractsManager_Upgrade_Harness {
+/// @notice Tests the `validate` function of the `OPContractsManager` contract.
+contract OPContractsManager_Validate_Test is OPContractsManager_TestInit {
+    /// @notice Tests that validate function returns validation results.
+    function test_validate_succeeds() public view {
```
I think this is generally ok, although since this is basically just a wrapper for the StandardValidator it means we potentially risk duplicating a ton of tests. I think it's okay.

smartcontracts commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -2083,3 +2101,134 @@ contract OPContractsManager_SetRC_Test is OPContractsManager_Upgrade_Harness {
+
+        // When allowFailure is false and validation fails, it should revert
+        // The error message contains validation failure codes
+        vm.expectRevert(
```
Meh this test will probably get annoying, maybe we can just vm.expectRevert without any error string (you will need to have semgrep ignore it).

smartcontracts commented on packages/contracts-bedrock/test/L1/OPContractsManager.t.sol:
```diff
@@ -2083,3 +2101,134 @@ contract OPContractsManager_SetRC_Test is OPContractsManager_Upgrade_Harness {
+    function test_blueprints_succeeds() public view {
+        IOPContractsManager.Blueprints memory blueprintAddresses = opcm.blueprints();
+
+        // Verify that blueprint addresses are set (non-zero)
```
Test will get stale, will need continuous updating, probably fine for now if we figure out how to integrate the AI to more regularly review and update this type of thing.

smartcontracts replied:
Recommend having the prompt try to identify and tag tests that could get stale so it has a way to find/review those tests later.

aliersh commented:
Closed: Tests are now stale due to contract changes in OPContractsManager. Will regenerate with updated prompt (v0.8.1) to reflect the latest implementation.
